File: TestBrotliDotNet/TestBrotliGZip/CompressedNetTest.cs

using System.Text;

namespace michele.natale.Tests;

using Compresses;
using Services;

public class CompressedNetTest
{
    private static string OriginalString =>
      File.ReadAllText("data.txt"); //5.294 MB

    public async static Task Start()
    {
        TestGzip();
        await TestGzipAsync();

        TestGzipFileStream();
        await TestGzipFileStreamAsync();

        await TestGzipStressAsync(); //Approx 135 MB


        TestBrotli();
        await TestBrotliAsync();

        TestBrotliStream();
        await TestBrotliStreamAsync();

        await TestBrotliStressAsync();//Approx 135 MB

        Console.WriteLine();
    }


    private static void TestGzip()
    {
        var message = Encoding.UTF8.GetBytes(OriginalString);

        //GZip Classic
        Console.WriteLine("GZip Classic");
        var compress = CompressedNet.CompressGZip(message);
        Console.WriteLine("Length of compressed bytes: " + compress.Length);

        var decompress = CompressedNet.DecompressGZip(compress);
        var ratio = Math.Round(compress.Length / (double)decompress.Length, 6);
        Console.WriteLine("Length of decompressed bytes: " + decompress.Length);
        Console.WriteLine($"ratio = {ratio} >> {Math.Round(100 * ratio, 5)}%");
        Console.WriteLine();
        if (!message.SequenceEqual(decompress)) throw new Exception();


        //GZip with CancellationToken
        Console.WriteLine("GZip with CancellationToken");
        var cts = new CancellationTokenSource();
        compress = CompressedNet.CompressGZip(message, cts.Token);
        Console.WriteLine("Length of compressed bytes: " + compress.Length);

        cts = new CancellationTokenSource();
        decompress = CompressedNet.DecompressGZip(compress, cts.Token);
        ratio = Math.Round(compress.Length / (double)decompress.Length, 6);
        Console.WriteLine("Length of decompressed bytes: " + decompress.Length);
        Console.WriteLine($"ratio = {ratio} >> {Math.Round(100 * ratio, 5)}%");
        Console.WriteLine();
        if (!message.SequenceEqual(decompress)) throw new Exception();

        Console.WriteLine();
    }

    private async static Task TestGzipAsync()
    {
        var message = Encoding.UTF8.GetBytes(OriginalString);

        //GZip Async 
        Console.WriteLine("GZip Async");
        var compress1 = await CompressedNet.CompressGZipAsync(message);
        Console.WriteLine("Length of compressed bytes: " + (compress1).Length);

        var decompress1 = await CompressedNet.DecompressGZipAsync(compress1);
        var ratio = Math.Round(compress1.Length / (double)decompress1.Length, 6);
        Console.WriteLine("Length of decompressed bytes: " + decompress1.Length);
        Console.WriteLine($"ratio = {ratio} >> {Math.Round(100 * ratio, 5)}%");
        Console.WriteLine();

        if (!message.SequenceEqual(decompress1)) throw new Exception();


        //GZip Async with CancellationToken
        Console.WriteLine("GZip Async with CancellationToken");
        var cts = new CancellationTokenSource();
        var compress2 = await CompressedNet.CompressGZipAsync(message, cts.Token);
        Console.WriteLine("Length of compressed string: " + compress2.Length);

        cts = new CancellationTokenSource();
        var decompress2 = await CompressedNet.DecompressGZipAsync(compress2, cts.Token);
        ratio = Math.Round(compress2.Length / (double)decompress2.Length, 6);
        Console.WriteLine("Length of decompressed string: " + decompress2.Length);
        Console.WriteLine($"ratio = {ratio} >> {Math.Round(100 * ratio, 5)}%");
        Console.WriteLine();
        if (!message.SequenceEqual(decompress2)) throw new Exception();

        Console.WriteLine();
    }

    private static void TestGzipFileStream()
    {
        string src = "data.txt", dest = "datacompress", destr = "datar.txt";
        File.Delete(dest); File.Delete(destr);

        //GZip Stream Classic
        Console.WriteLine("GZip Stream Classic");
        CompressedNet.CompressGZip(src, dest);
        Console.WriteLine("Length of compressed bytes: " + ServicesCompress.FileSize(dest));

        CompressedNet.DecompressGZip(dest, destr);
        Console.WriteLine("Length of decompressed bytes: " + ServicesCompress.FileSize(destr));
        var ratio = Math.Round(ServicesCompress.FileSize(dest) / (double)ServicesCompress.FileSize(destr), 6);
        Console.WriteLine($"ratio = {ratio} >> {Math.Round(100 * ratio, 5)}%");
        Console.WriteLine();
        if (!ServicesCompress.FileEquals(src, destr)) throw new Exception();


        //GZip Stream with CancellationToken
        Console.WriteLine("GZip Stream with CancellationToken");
        var cts = new CancellationTokenSource();
        CompressedNet.CompressGZip(src, dest, cts.Token);
        Console.WriteLine("Length of compressed bytes: " + ServicesCompress.FileSize(dest));

        cts = new CancellationTokenSource();
        CompressedNet.DecompressGZip(dest, destr, cts.Token);
        Console.WriteLine("Length of decompressed bytes: " + ServicesCompress.FileSize(destr));
        ratio = Math.Round(ServicesCompress.FileSize(dest) / (double)ServicesCompress.FileSize(destr), 6);
        Console.WriteLine($"ratio = {ratio} >> {Math.Round(100 * ratio, 5)}%");
        Console.WriteLine();
        if (!ServicesCompress.FileEquals(src, destr)) throw new Exception();

        Console.WriteLine();
    }

    private static async Task TestGzipFileStreamAsync()
    {
        string src = "data.txt", dest = "datacompress", destr = "datar.txt";
        File.Delete(dest); File.Delete(destr);

        //GZip Stream Async
        Console.WriteLine("GZip Stream Async");
        await CompressedNet.CompressGZipAsync(src, dest);
        Console.WriteLine("Length of compressed bytes: " + await ServicesCompress.FileSizeLongAsync(dest));

        await CompressedNet.DecompressGZipAsync(dest, destr);
        Console.WriteLine("Length of decompressed bytes: " + await ServicesCompress.FileSizeLongAsync(destr));
        var ratio = Math.Round(await ServicesCompress.FileSizeLongAsync(dest) / (double)await ServicesCompress.FileSizeLongAsync(destr), 6);
        Console.WriteLine($"ratio = {ratio} >> {Math.Round(100 * ratio, 5)}%");
        Console.WriteLine();
        if (!ServicesCompress.FileEquals(src, destr)) throw new Exception();

        File.Delete(dest); File.Delete(destr);

        //GZip Stream Async with CancellationToken
        Console.WriteLine("GZip Stream Async with CancellationToken");

        var cts = new CancellationTokenSource();
        await CompressedNet.CompressGZipAsync(src, dest, cts.Token);
        Console.WriteLine("Length of compressed bytes: " + await ServicesCompress.FileSizeLongAsync(dest));

        cts = new CancellationTokenSource();
        await CompressedNet.DecompressGZipAsync(dest, destr, cts.Token);
        Console.WriteLine("Length of decompressed bytes: " + await ServicesCompress.FileSizeLongAsync(destr));
        ratio = Math.Round(await ServicesCompress.FileSizeLongAsync(dest) / (double)await ServicesCompress.FileSizeLongAsync(destr), 6);
        Console.WriteLine($"ratio = {ratio} >> {Math.Round(100 * ratio, 5)}%");
        Console.WriteLine();
        if (!ServicesCompress.FileEquals(src, destr)) throw new Exception();

        Console.WriteLine();
    }

    private async static Task TestGzipStressAsync()
    {
        Console.WriteLine("GZip Stream Async Stresstest");
        string src = "data_rng.txt", dest = "datacompress", destr = "datar.txt";
        File.Delete(dest); File.Delete(destr);

        var bytes = Encoding.UTF8.GetBytes(OriginalString);
        await ServicesCompress.MultBytesInFileAsync(src, bytes, 25);
        await CompressedNet.CompressGZipAsync(src, dest);
        Console.WriteLine("Length of compressed bytes: " + await ServicesCompress.FileSizeLongAsync(dest));

        await CompressedNet.DecompressGZipAsync(dest, destr);
        Console.WriteLine("Length of decompressed bytes: " + await ServicesCompress.FileSizeLongAsync(destr));
        var ratio = Math.Round(await ServicesCompress.FileSizeLongAsync(dest) / (double)await ServicesCompress.FileSizeLongAsync(destr), 12);
        Console.WriteLine($"ratio = {ratio.ToString("F12")} >> {Math.Round(100 * ratio, 8)}%");
        Console.WriteLine();
        if (!ServicesCompress.FileEquals(src, destr)) throw new Exception();

        File.Delete(src); File.Delete(dest); File.Delete(destr);

        Console.WriteLine();
    }

    private static void TestBrotli()
    {
        var message = Encoding.UTF8.GetBytes(OriginalString);

        //Brotli Classic
        Console.WriteLine("Brotli Classic");
        var compress = CompressedNet.CompressBrotli(message);
        Console.WriteLine("Length of compressed bytes: " + compress.Length);

        var decompress = CompressedNet.DecompressBrotli(compress);
        Console.WriteLine("Length of decompressed bytes: " + decompress.Length);
        var ratio = Math.Round(compress.Length / (double)decompress.Length, 6);
        Console.WriteLine($"ratio = {ratio} >> {Math.Round(100 * ratio, 5)}%");
        Console.WriteLine();
        if (!message.SequenceEqual(decompress)) throw new Exception();


        //Brotli with CancellationToken
        Console.WriteLine("Brotli with CancellationToken");
        var cts = new CancellationTokenSource();
        compress = CompressedNet.CompressBrotli(message, cts.Token);
        Console.WriteLine("Length of compressed bytes: " + compress.Length);

        cts = new CancellationTokenSource();
        decompress = CompressedNet.DecompressBrotli(compress, cts.Token);
        Console.WriteLine("Length of decompressed bytes: " + decompress.Length);
        ratio = Math.Round(compress.Length / (double)decompress.Length, 6);
        Console.WriteLine($"ratio = {ratio} >> {Math.Round(100 * ratio, 5)}%");
        Console.WriteLine();
        if (!message.SequenceEqual(decompress)) throw new Exception();


        //Brotli Try
        Console.WriteLine("Brotli Try");
        int quality = 4, window = 22;
        if (!CompressedNet.TryCompressBrotli(
          message, out compress, out var written, quality, window))
            throw new Exception();
        Console.WriteLine("Length of compressed bytes: " + written);

        if (!CompressedNet.TryDecompressBrotli(compress, out decompress, out written))
            throw new Exception();
        Console.WriteLine("Length of decompressed bytes: " + written);
        ratio = Math.Round(compress.Length / (double)decompress.Length, 6);
        Console.WriteLine($"ratio = {ratio} >> {Math.Round(100 * ratio, 5)}%");
        Console.WriteLine();
        if (!message.SequenceEqual(decompress)) throw new Exception();


        //Brotli Try
        Console.WriteLine("Brotli Try with CancellationToken");
        cts = new CancellationTokenSource();
        if (!CompressedNet.TryCompressBrotli(
          message, cts.Token, out compress, out written, quality, window))
            throw new Exception();
        Console.WriteLine("Length of compressed bytes: " + written);

        if (!CompressedNet.TryDecompressBrotli(
          compress, cts.Token, out decompress, out written))
            throw new Exception();
        Console.WriteLine("Length of decompressed bytes: " + written);
        ratio = Math.Round(compress.Length / (double)decompress.Length, 6);
        Console.WriteLine($"ratio = {ratio} >> {Math.Round(100 * ratio, 5)}%");
        Console.WriteLine();
        if (!message.SequenceEqual(decompress)) throw new Exception();

        Console.WriteLine();
    }

    private async static Task TestBrotliAsync()
    {
        var message = Encoding.UTF8.GetBytes(OriginalString);

        //Brotli Async 
        Console.WriteLine("Brotli Async");
        var compress1 = await CompressedNet.CompressBrotliAsync(message);
        Console.WriteLine("Length of compressed bytes: " + compress1.Length);

        var decompress1 = await CompressedNet.DecompressBrotliAsync(compress1);
        Console.WriteLine("Length of decompressed bytes: " + decompress1.Length);
        var ratio = Math.Round(compress1.Length / (double)decompress1.Length, 6);
        Console.WriteLine($"ratio = {ratio} >> {Math.Round(100 * ratio, 5)}%");
        Console.WriteLine();
        if (!message.SequenceEqual(decompress1)) throw new Exception();


        //Brotli Async with CancellationToken
        Console.WriteLine("Brotli Async with CancellationToken");
        var cts = new CancellationTokenSource();
        var compress2 = await CompressedNet.CompressBrotliAsync(message, cts.Token);
        Console.WriteLine("Length of compressed string: " + compress2.Length);

        cts = new CancellationTokenSource();
        var decompress2 = await CompressedNet.DecompressBrotliAsync(compress2, cts.Token);
        Console.WriteLine("Length of decompressed string: " + decompress2.Length);
        ratio = Math.Round(compress2.Length / (double)decompress2.Length, 6);
        Console.WriteLine($"ratio = {ratio} >> {Math.Round(100 * ratio, 5)}%");
        Console.WriteLine();
        if (!message.SequenceEqual(decompress2)) throw new Exception();

        Console.WriteLine();
    }

    private static void TestBrotliStream()
    {
        string src = "data.txt", dest = "datacompress", destr = "datar.txt";
        File.Delete(dest); File.Delete(destr);

        //Brotli Stream Classic
        Console.WriteLine("Brotli Stream Classic");
        CompressedNet.CompressBrotli(src, dest);
        Console.WriteLine("Length of compressed bytes: " + ServicesCompress.FileSize(dest));

        CompressedNet.DecompressBrotli(dest, destr);
        Console.WriteLine("Length of decompressed bytes: " + ServicesCompress.FileSize(destr));
        var ratio = Math.Round(ServicesCompress.FileSize(dest) / (double)ServicesCompress.FileSize(destr), 6);
        Console.WriteLine($"ratio = {ratio} >> {Math.Round(100 * ratio, 5)}%");
        Console.WriteLine();
        if (!ServicesCompress.FileEquals(src, destr)) throw new Exception();


        //Brotli Stream with CancellationToken
        Console.WriteLine("Brotli Stream with CancellationToken");
        var cts = new CancellationTokenSource();
        CompressedNet.CompressBrotli(src, dest, cts.Token);
        Console.WriteLine("Length of compressed bytes: " + ServicesCompress.FileSize(dest));

        cts = new CancellationTokenSource();
        CompressedNet.DecompressBrotli(dest, destr, cts.Token);
        Console.WriteLine("Length of decompressed bytes: " + ServicesCompress.FileSize(destr));
        ratio = Math.Round(ServicesCompress.FileSize(dest) / (double)ServicesCompress.FileSize(destr), 6);
        Console.WriteLine($"ratio = {ratio} >> {Math.Round(100 * ratio, 5)}%");
        Console.WriteLine();
        if (!ServicesCompress.FileEquals(src, destr)) throw new Exception();

        Console.WriteLine();
    }

    private async static Task TestBrotliStreamAsync()
    {
        string src = "data.txt", dest = "datacompress", destr = "datar.txt";
        File.Delete(dest); File.Delete(destr);

        //Brotli Stream Async
        Console.WriteLine("Brotli Stream Async");
        await CompressedNet.CompressBrotliAsync(src, dest);
        Console.WriteLine("Length of compressed bytes: " + await ServicesCompress.FileSizeLongAsync(dest));

        await CompressedNet.DecompressBrotliAsync(dest, destr);
        Console.WriteLine("Length of decompressed bytes: " + await ServicesCompress.FileSizeLongAsync(destr));
        var ratio = Math.Round(await ServicesCompress.FileSizeLongAsync(dest) / (double)await ServicesCompress.FileSizeLongAsync(destr), 6);
        Console.WriteLine($"ratio = {ratio} >> {Math.Round(100 * ratio, 5)}%");
        Console.WriteLine();
        if (!ServicesCompress.FileEquals(src, destr)) throw new Exception();

        File.Delete(dest); File.Delete(destr);

        //Brotli Stream Async with CancellationToken
        Console.WriteLine("Brotli Stream Async with CancellationToken");
        var cts = new CancellationTokenSource();
        await CompressedNet.CompressBrotliAsync(src, dest, cts.Token);
        Console.WriteLine("Length of compressed bytes: " + await ServicesCompress.FileSizeLongAsync(dest));

        cts = new CancellationTokenSource();
        await CompressedNet.DecompressBrotliAsync(dest, destr, cts.Token);
        Console.WriteLine("Length of decompressed bytes: " + await ServicesCompress.FileSizeLongAsync(destr));
        ratio = Math.Round(await ServicesCompress.FileSizeLongAsync(dest) / (double)await ServicesCompress.FileSizeLongAsync(destr), 6);
        Console.WriteLine($"ratio = {ratio} >> {Math.Round(100 * ratio, 5)}%");
        Console.WriteLine();
        if (!ServicesCompress.FileEquals(src, destr)) throw new Exception();

        Console.WriteLine();
    }


    private async static Task TestBrotliStressAsync()
    {
        Console.WriteLine("Brotli Stream Async Stresstest");
        string src = "data_rng.txt", dest = "datacompress", destr = "datar.txt";
        File.Delete(dest); File.Delete(destr);

        var bytes = Encoding.UTF8.GetBytes(OriginalString);
        await ServicesCompress.MultBytesInFileAsync(src, bytes, 25);
        await CompressedNet.CompressBrotliAsync(src, dest);
        Console.WriteLine("Length of compressed bytes: " + await ServicesCompress.FileSizeLongAsync(dest));

        await CompressedNet.DecompressBrotliAsync(dest, destr);
        Console.WriteLine("Length of decompressed bytes: " + await ServicesCompress.FileSizeLongAsync(destr));
        var ratio = Math.Round(await ServicesCompress.FileSizeLongAsync(dest) / (double)await ServicesCompress.FileSizeLongAsync(destr), 12);
        Console.WriteLine($"ratio = {ratio.ToString("F12")} >> {Math.Round(100 * ratio, 8)}%");
        Console.WriteLine();
        if (!await ServicesCompress.FileEqualsAsync(src, destr)) throw new Exception();

        File.Delete(src); File.Delete(dest); File.Delete(destr);

        Console.WriteLine();
    }

}

File: TestBrotliDotNet/TestBrotliGZipVB/CompressedNetTest.vb


Imports System.IO
Imports System.Text
Imports System.Threading
Imports michele.natale.Services
Imports michele.natale.Compresses


Namespace michele.natale.Tests
  Public Class CompressedNetTest

    Private Shared ReadOnly Property OriginalString As String
      Get
        Return File.ReadAllText("data.txt") '5.294 MB
      End Get
    End Property

    Public Shared Async Function Start() As Task
      TestGzip()
      Await TestGzipAsync()

      TestGzipFileStream()
      Await TestGzipFileStreamAsync()

      'Await TestGzipStressAsync() 'Approx 135 MB

      TestBrotli()
      Await TestBrotliAsync()

      TestBrotliStream()
      Await TestBrotliStreamAsync()

      'Await TestBrotliStressAsync() 'Approx 135 MB

      Console.WriteLine()
    End Function

    Private Shared Sub TestGzip()
      Dim message = Encoding.UTF8.GetBytes(OriginalString)

      'GZip Classic
      Console.WriteLine("GZip Classic")
      Dim compress = CompressedNet.CompressGZip(message)
      Console.WriteLine("Length of compressed bytes: " & compress.Length)

      Dim decompress = CompressedNet.DecompressGZip(compress)
      Console.WriteLine("Length of decompressed bytes: " & decompress.Length)
      Console.WriteLine()
      If Not message.SequenceEqual(decompress) Then Throw New Exception()

      'GZip with CancellationToken
      Console.WriteLine("GZip with CancellationToken")
      Dim cts = New CancellationTokenSource()
      compress = CompressedNet.CompressGZip(message, cts.Token)
      Console.WriteLine("Length of compressed bytes: " & compress.Length)

      cts = New CancellationTokenSource()
      decompress = CompressedNet.DecompressGZip(compress, cts.Token)
      Console.WriteLine("Length of decompressed bytes: " & decompress.Length)
      Console.WriteLine()
      If Not message.SequenceEqual(decompress) Then Throw New Exception()

      Console.WriteLine()
    End Sub

    Private Shared Async Function TestGzipAsync() As Task
      Dim message = Encoding.UTF8.GetBytes(OriginalString)

      'GZip Async 
      Console.WriteLine("GZip Async")
      Dim compress1 = Await CompressedNet.CompressGZipAsync(message)
      Console.WriteLine("Length of compressed bytes: " & compress1.Length)

      Dim decompress1 = Await CompressedNet.DecompressGZipAsync(compress1)
      Console.WriteLine("Length of decompressed bytes: " & decompress1.Length)
      Console.WriteLine()

      If Not message.SequenceEqual(decompress1) Then Throw New Exception()

      'GZip Async with CancellationToken
      Console.WriteLine("GZip Async with CancellationToken")
      Dim cts = New CancellationTokenSource()
      Dim compress2 = Await CompressedNet.CompressGZipAsync(message, cts.Token)
      Console.WriteLine("Length of compressed string: " & compress2.Length)

      cts = New CancellationTokenSource()
      Dim decompress2 = Await CompressedNet.DecompressGZipAsync(compress2, cts.Token)
      Console.WriteLine("Length of decompressed string: " & decompress2.Length)
      Console.WriteLine()
      If Not message.SequenceEqual(decompress2) Then Throw New Exception()

      Console.WriteLine()
    End Function

    Private Shared Sub TestGzipFileStream()
      Dim src As String = "data.txt"
      Dim dest As String = "datacompress"
      Dim destr As String = "datar.txt"
      File.Delete(dest)
      File.Delete(destr)

      'GZip Stream Classic
      Console.WriteLine("GZip Stream Classic")
      CompressedNet.CompressGZip(src, dest)
      Console.WriteLine("Length of compressed bytes: " & ServicesCompress.FileSize(dest))

      CompressedNet.DecompressGZip(dest, destr)
      Console.WriteLine("Length of decompressed bytes: " & ServicesCompress.FileSize(destr))
      Console.WriteLine()
      If Not ServicesCompress.FileEquals(src, destr) Then Throw New Exception()

      'GZip Stream with CancellationToken
      Console.WriteLine("GZip Stream with CancellationToken")
      Dim cts = New CancellationTokenSource()
      CompressedNet.CompressGZip(src, dest, cts.Token)
      Console.WriteLine("Length of compressed bytes: " & ServicesCompress.FileSize(dest))

      cts = New CancellationTokenSource()
      CompressedNet.DecompressGZip(dest, destr, cts.Token)
      Console.WriteLine("Length of decompressed bytes: " & ServicesCompress.FileSize(destr))
      Console.WriteLine()
      If Not ServicesCompress.FileEquals(src, destr) Then Throw New Exception()

      Console.WriteLine()
    End Sub


    Private Shared Async Function TestGzipFileStreamAsync() As Task
      Dim src As String = "data.txt"
      Dim dest As String = "datacompress"
      Dim destr As String = "datar.txt"
      File.Delete(dest) : File.Delete(destr)

      'GZip Stream Async
      Console.WriteLine("GZip Stream Async")
      Await CompressedNet.CompressGZipAsync(src, dest)
      Console.WriteLine("Length of compressed bytes: " & ServicesCompress.FileSize(dest))

      Await CompressedNet.DecompressGZipAsync(dest, destr)
      Console.WriteLine("Length of decompressed bytes: " & ServicesCompress.FileSize(destr))
      Console.WriteLine()
      If Not ServicesCompress.FileEquals(src, destr) Then Throw New Exception()

      File.Delete(dest) : File.Delete(destr)

      'GZip Stream Async with CancellationToken
      Console.WriteLine("GZip Stream Async with CancellationToken")

      Dim cts As New CancellationTokenSource()
      Await CompressedNet.CompressGZipAsync(src, dest, cts.Token)
      Console.WriteLine("Length of compressed bytes: " & ServicesCompress.FileSize(dest))

      cts = New CancellationTokenSource()
      Await CompressedNet.DecompressGZipAsync(dest, destr, cts.Token)
      Console.WriteLine("Length of decompressed bytes: " & ServicesCompress.FileSize(destr))
      Console.WriteLine()
      If Not ServicesCompress.FileEquals(src, destr) Then Throw New Exception()

      Console.WriteLine()
    End Function

    Private Shared Sub TestBrotli()
      Dim message As Byte() = Encoding.UTF8.GetBytes(OriginalString)

      'Brotli Classic
      Console.WriteLine("Brotli Classic")
      Dim compress As Byte() = CompressedNet.CompressBrotli(message)
      Console.WriteLine("Length of compressed bytes: " & compress.Length)

      Dim decompress As Byte() = CompressedNet.DecompressBrotli(compress)
      Console.WriteLine("Length of decompressed bytes: " & decompress.Length)
      Console.WriteLine()
      If Not message.SequenceEqual(decompress) Then Throw New Exception()

      'Brotli with CancellationToken
      Console.WriteLine("Brotli with CancellationToken")
      Dim cts As New CancellationTokenSource()
      compress = CompressedNet.CompressBrotli(message, cts.Token)
      Console.WriteLine("Length of compressed bytes: " & compress.Length)

      cts = New CancellationTokenSource()
      decompress = CompressedNet.DecompressBrotli(compress, cts.Token)
      Console.WriteLine("Length of decompressed bytes: " & decompress.Length)
      Console.WriteLine()
      If Not message.SequenceEqual(decompress) Then Throw New Exception()

      'Brotli Try
      Console.WriteLine("Brotli Try")
      Dim quality As Integer = 4
      Dim window As Integer = 22
      Dim written As Integer
      If Not CompressedNet.TryCompressBrotli(message, compress, written, quality, window) Then
        Throw New Exception()
      End If
      Console.WriteLine("Length of compressed bytes: " & written)

      If Not CompressedNet.TryDecompressBrotli(compress, decompress, written) Then
        Throw New Exception()
      End If
      Console.WriteLine("Length of decompressed bytes: " & written)
      Console.WriteLine()
      If Not message.SequenceEqual(decompress) Then Throw New Exception()

      'Brotli Try with CancellationToken
      Console.WriteLine("Brotli Try with CancellationToken")
      cts = New CancellationTokenSource()
      If Not CompressedNet.TryCompressBrotli(message, cts.Token, compress, written, quality, window) Then
        Throw New Exception()
      End If
      Console.WriteLine("Length of compressed bytes: " & written)

      If Not CompressedNet.TryDecompressBrotli(compress, cts.Token, decompress, written) Then
        Throw New Exception()
      End If
      Console.WriteLine("Length of decompressed bytes: " & written)
      Console.WriteLine()
      If Not message.SequenceEqual(decompress) Then Throw New Exception()

      Console.WriteLine()
    End Sub

    Private Shared Async Function TestBrotliAsync() As Task
      Dim message = Encoding.UTF8.GetBytes(OriginalString)

      'Brotli Async 
      Console.WriteLine("Brotli Async")
      Dim compress1 = Await CompressedNet.CompressBrotliAsync(message)
      Console.WriteLine("Length of compressed bytes: " & compress1.Length)

      Dim decompress1 = Await CompressedNet.DecompressBrotliAsync(compress1)
      Console.WriteLine("Length of decompressed bytes: " & decompress1.Length)
      Console.WriteLine()
      If Not message.SequenceEqual(decompress1) Then Throw New Exception()

      'Brotli Async with CancellationToken
      Console.WriteLine("Brotli Async with CancellationToken")
      Dim cts = New CancellationTokenSource()
      Dim compress2 = Await CompressedNet.CompressBrotliAsync(message, cts.Token)
      Console.WriteLine("Length of compressed string: " & compress2.Length)

      cts = New CancellationTokenSource()
      Dim decompress2 = Await CompressedNet.DecompressBrotliAsync(compress2, cts.Token)
      Console.WriteLine("Length of decompressed string: " & decompress2.Length)
      Console.WriteLine()
      If Not message.SequenceEqual(decompress2) Then Throw New Exception()

      Console.WriteLine()
    End Function

    Private Shared Sub TestBrotliStream()
      Dim src As String = "data.txt"
      Dim dest As String = "datacompress"
      Dim destr As String = "datar.txt"
      File.Delete(dest)
      File.Delete(destr)

      'Brotli Stream Classic
      Console.WriteLine("Brotli Stream Classic")
      CompressedNet.CompressBrotli(src, dest)
      Console.WriteLine("Length of compressed bytes: " & ServicesCompress.FileSize(dest))

      CompressedNet.DecompressBrotli(dest, destr)
      Console.WriteLine("Length of decompressed bytes: " & ServicesCompress.FileSize(destr))
      Console.WriteLine()
      If Not ServicesCompress.FileEquals(src, destr) Then Throw New Exception()

      'Brotli Stream with CancellationToken
      Console.WriteLine("Brotli Stream with CancellationToken")
      Dim cts = New CancellationTokenSource()
      CompressedNet.CompressBrotli(src, dest, cts.Token)
      Console.WriteLine("Length of compressed bytes: " & ServicesCompress.FileSize(dest))

      cts = New CancellationTokenSource()
      CompressedNet.DecompressBrotli(dest, destr, cts.Token)
      Console.WriteLine("Length of decompressed bytes: " & ServicesCompress.FileSize(destr))
      Console.WriteLine()
      If Not ServicesCompress.FileEquals(src, destr) Then Throw New Exception()

      Console.WriteLine()
    End Sub

    Private Shared Async Function TestBrotliStreamAsync() As Task
      Dim src As String = "data.txt"
      Dim destr As String = "datar.txt"
      Dim dest As String = "datacompress"
      File.Delete(dest) : File.Delete(destr)

      'Brotli Stream Async
      Console.WriteLine("Brotli Stream Async")
      Await CompressedNet.CompressBrotliAsync(src, dest)
      Console.WriteLine("Length of compressed bytes: " & ServicesCompress.FileSize(dest))

      Await CompressedNet.DecompressBrotliAsync(dest, destr)
      Console.WriteLine("Length of decompressed bytes: " & ServicesCompress.FileSize(destr))
      Console.WriteLine()
      If Not ServicesCompress.FileEquals(src, destr) Then Throw New Exception()

      File.Delete(dest)
      File.Delete(destr)

      'Brotli Stream Async with CancellationToken
      Console.WriteLine("Brotli Stream Async with CancellationToken")
      Dim cts As New CancellationTokenSource()
      Await CompressedNet.CompressBrotliAsync(src, dest, cts.Token)
      Console.WriteLine("Length of compressed bytes: " & ServicesCompress.FileSize(dest))

      cts = New CancellationTokenSource()
      Await CompressedNet.DecompressBrotliAsync(dest, destr, cts.Token)
      Console.WriteLine("Length of decompressed bytes: " & ServicesCompress.FileSize(destr))
      Console.WriteLine()
      If Not ServicesCompress.FileEquals(src, destr) Then Throw New Exception()

      Console.WriteLine()
    End Function
  End Class
End Namespace
